### PR TITLE
Ds 1138/docstrings

### LINKED ISSUE #83: Add docstrings to readthedocs

### PR DESCRIPTION
**Pull request by gsganden**
Move most docs into readthedocs, organize them, add docstrings.

I would suggest reviewing [the docs themselves](https://creevey.readthedocs.io/en/ds-1138-docstrings/) before looking at the implementation.

I also renamed `pipelines/core.py` to `pipelines/pipelines.py` to match the naming convention used elsewhere in the library.

Lots of files have been changed because I added module docstrings, which are pretty useful in readthedocs, and had to tweak some of the existing docstrings to work properly with autodoc.

Contributors from outside ShopRunner should feel free to submit a PR without having completed all of the items below. See [CONTRIBUTING.md](https://github.com/ShopRunner/creevey/blob/master/CONTRIBUTING.md) for additional information. 

Closes #83 

**General Pull Request Checklist**

 - [x] Pull request has been made against the `wip` branch (or from `wip` into `master`).
 - [x] Pull request includes a description of the change and the reason behind it.
 - [x] Pull request [uses keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to close relevant [issues](https://github.com/ShopRunner/creevey/issues).
 - [x] Pull request includes unit tests for any bug fixes and new functionality.
 - [x] `README.md` and other docs have been updated as needed.
 - [x] `./.ci/test.sh` passes locally.
 
The maintainer will complete the following steps for external contributions.

**Additional Items for ShopRunner Contributors**

 - [x] `CHANGELOG.md` has been updated.
 - [x] `_version.py` has been updated.

### CODE REVIEW

File: CONDUCT.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Contributor Covenant Code of Conduct
```
Comment (gsganden) [inline CONDUCT.md:1]: The verbose name annoyed me

File: README.md [hunk 2]
```diff
@@ -2,155 +2,8 @@
 
 ![](https://images.pottermore.com/bxd3o8b291gf/22qh5bCcA0g28OeKCwgwgE/70be84ace5da257fbd54d1ca0c06972c/ColinCreevey_WB_F2_ColinHoldingCamera_Still_080615_Land.jpg?w=320&h=320&fit=thumb&f=left&q=85)
 
-**Creevey provides a simple framework for batch file processing pipelines that handles threading, piping, logging, and (optionally) skipping existing files for you.** It is designed for IO-bound batch workflows that involve reading files into memory, doing some processing on their contents, and writing out the results. **Creevey also provides predefined, extensible pipelines and reusable pipeline components, particularly for image processing.** 
+Creevey is a file processing framework. It is designed for IO-bound workflows that involve reading files into memory, doing some processing on their contents, and writing out the results. It makes running those workflows faster and more reliable by parallelizing them across files, optional skipping files that have already been processed, handling errors, and keeping organized records of what was done.
```
Comment (gsganden) [inline README.md:5]: The README now refers to the docs for details so that we have less to maintain in two places.

File: docs/contrib.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+.. _contrib:
+
+.. mdinclude:: ../CONTRIBUTING.md
+
+
+.. mdinclude:: ../CONDUCT.md
```
Comment (jscottcronin) [inline docs/contrib.rst:6]: Any desire to do separate .rst files for contributing and code of conduct? Read the docs treats as a single thing which feels odd. If you click on `Contributing` in the TOC, it opens both `Contributing` and `Conduct` as a single file.

File: creevey/constants.py [hunk 4]
```diff
@@ -1,4 +1,6 @@
+"""Constants"""
 from pathlib import Path
 from typing import Union
 
 PathOrStr = Union[Path, str]
+"""Path or string type"""
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/constants.py:6]: It adds an entry to https://creevey.readthedocs.io/en/ds-1138-docstrings/contents.html#module-creevey.constants

File: creevey/ops/helpers/report.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-"""Code for reporting information calculating during processing"""
+"""Code for reporting information calculating during processing."""
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/ops/helpers/report.py:1]: > Some module docstrings end in period some do not. Is that important / bothersome?

Not to me

> rendering seems a little odd:

It reflects the package structure: the module is `creevey/ops/helpers/report.py`

File: creevey/ops/image/transforms.py [hunk 6]
```diff
@@ -31,10 +32,6 @@ def resize(
         resized so that it has this length along its smaller spatial
         dimension while preseving aspect ratio as closely as possible.
         Exactly one of `shape` and `min_dim` must be `None`.
-
-    Returns
-    -------
-    NumPy array with specified shape
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/ops/image/transforms.py:37]: Yes, but that information is already in the type hint, so these lines are redundant.

File: creevey/pipelines/image.py [hunk 7]
```diff
@@ -1,7 +1,10 @@
+"""Image-processing pipelines"""
 from creevey import Pipeline
 from creevey.load_funcs.image import load_image_from_url
 from creevey.write_funcs.image import write_image
 
+
 download_image_pipeline = Pipeline(
     load_func=load_image_from_url, ops=[], write_func=write_image
 )
+"""Basic pipeline for downloading images"""
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/pipelines/image.py:10]: Yes, I think so. Do you have suggestions for improving it?

File: creevey/pipelines/pipelines.py [hunk 8]
```diff
@@ -148,7 +148,7 @@ def run(self, *args, **kwargs):
         """
```
[the remark this answers is not in the document]
Comment (gsganden) [reply creevey/pipelines/pipelines.py:148]: We're only keeping `run` for backwards compatibility; I intend for people to call pipelines directly and have updated the quickstart examples accordingly.

File: docs/contents.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
+Contents
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/contents.rst:1]: Yeah, I like that.

File: docs/contents.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+.. automodule:: creevey.pipelines.pipelines
+   :special-members: __call__
+   :members:
+
+image
+^^^^^
+
+.. automodule:: creevey.pipelines.image
+   :members:
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/contents.rst:20]: I don't get the confusion. It's telling us that `pipelines/image.py` is for image-processing pipelines. Currently it contains `download_image_pipeline`, which is a basic pipeline for downloading images.

File: docs/contents.rst [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+load_funcs
+----------
+
+load_funcs
+^^^^^^^^^^
+
+.. automodule:: creevey.load_funcs.load_funcs
+   :members:
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/contents.rst:29]: It looks like the `#` in the URL interfered with the formatting. I'll just take out the URL.

File: docs/quickstart.rst [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+       inpaths=image_urls,
+       path_func=keep_filename_png_in_cwd,
+       n_jobs=10,
+       skip_existing=True,
+       exceptions_to_catch=AttributeError,
+   )
+
+``trim_resize_pipeline(...)`` returns a "run report:" a Pandas DataFrame with each input path as its index and columns indicating the corresponding output path ("outpath"), whether processing was skipped because a file already existed at the output path ("skipped_existing"), whether processing failed due to an exception in ``exceptions_to_catch`` ("exception_handled"), and a timestamp indicating when processing completed ("time_finished"):
+
+If you are doing exploratory work and don't really care about the errors right now, you can pass ``exceptions_to_catch=BaseException`` to the pipeline call above, and creevey will simply catch all Exceptions, downloading as many images as it can.
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/quickstart.rst]: Fixed

File: docs/quickstart.rst [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+.. image:: ./images/run_report_image.png
+   :target: ./images/run_report_image.png
+   :alt:
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/quickstart.rst:51]: It's a unix timestamp. I'm going to change that.

File: docs/quickstart.rst [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+Running a ``Pipeline``
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A ``Pipeline`` object takes the following arguments when called:
+
+
+#. An iterable ``inpaths`` of input paths (a list of image URLs in this example).
+#. A function ``outpath_func`` for transforming each path in ``inpaths`` into a corresponding output path. In this example, ``keep_filename_png_in_cwd`` uses the filename from the URL but gives it a PNG extension and places it in the current working directory.
+#. The number ``n_jobs`` of threads to run (10 in this example).
```
[the remark this answers is not in the document]
Comment (gsganden) [reply docs/quickstart.rst:81]: I think it depends too much on the task and the environment to make a general recommendation.

File: docs/contrib.rst (hunk 3 above)
Comment (gsganden) [reply docs/contrib.rst:6]: Fixed, thanks

**Pull request merged**